torch/csrc/jit/frontend/sugared_value.cpp

```diff
@@ -625,6 +625,13 @@ std::shared_ptr<SugaredValue> ClassValue::attr(
     const SourceRange& loc,
     Function& m,
     const std::string& field) {
+  // Allow import_source.cpp to resolve calls to a submodule's
+  // hooks. Edge case because normally you wouldn't allow a module to
+  // call functions of a submodule
+  if (Function* hook = type_->findHook(field)) {
+    return std::make_shared<FunctionValue>(hook);
+  }
```

Comment:
I don't understand the reason this is needed. Under what circumstances are hooks accessed by name?

Reply:
It's because we're calling hooks as functions of submodules instead of methods. From how the serialization is currently implemented, when trying to resolve the function in the graph it see's that the function is stored as an attribute on the submodule's class type and it tried to resolve it. Here's a paste that might help: https://www.internalfb.com/intern/paste/P160808038/

Reply:
> From how the serialization is currently implemented

By this, do you mean that it is a consequence of serializing hooks as if they were methods (and probably using the class name as part of the prefix for the hook function when compiling it)?. We are essentially baking hooks into the type of a module, so this isn't wrong, but I'm planning to add static methods to classes as part of improving TorchBind and that *might* conflict with this. However, I think it's valid to say that hooks are part of the type but not part of the instance (just like static methods), and so if you have a static method named `foo` and you try to attach and script a hook with the same name, compilation will fail (due to changes you made in the first PR in this stack). 

I think this is another thing I'd like to ask for more opinions on. Out of curiosity (👀 ), do you have any idea how much work it would be to change this PR to serialize hooks as separate functions instead of as part of the class definition for a module?